### DISCUSSION TITLE
TypeScript typing

### DISCUSSION THREAD

**Discussion by mobsense** [sensedeep/dynamodb-onetable#471] [Polls]
I'm starting the process to convert the code from JS to TS.

But the current approach of dynamic typing of models leads to very complex types and a poor intellisense experience.

Very early on, we made the decision to use dynamic typing using TS magic rather than a build time tool. However, we've been hitting the limits of this magic (and our skills) and the result is pretty complex.

A build time tool to generate types for models could lead to:

* Much simpler type definitions
* More complete types that support all the OneTable model field properties

What do you think?

**Comment by mobsense** [comment]
Feel free to make other suggestions for improving OneTable as we do this major translation.

**Comment by rjmackay** [comment]
Is there a branch for TS support that we could contribute to? I was just playing with this locally

**Reply by rjmackay** [reply]
I would also probably vote for just keeping "TS magic". But I suspect there are ways to make it a little cleaner, and I find this stuff is much easier to do if you're writing pure TS.

**Reply by mobsense** [reply]
Agree. We are converting packages we use over to TS one by one and this package is getting close to the top to be converted.

However, the intellisense really leaves a lot to be desired.

**Reply by rjmackay** [reply]
Are there PRs I could look at in other repos where you've converted?
Would you be open to me starting a branch for onetable if I can make time?

**Reply by mobsense** [reply]
Unfortunately those packages are internal and not released.

What I'd love to do is start a prototype that will DEFINITELY be thrown away to explore some ideas. i.e. create a pull request and let the code fly. See what looks like a good direction, then we start afresh using that data to guide us.

The key questions I've got are:

* Do we keep the current code structure (files, core functions etc)
* What are the APIs where we've done too much duck-typing and other JS dirty tricks to map cleanly onto TS
* What do the *.d.ts files look like at the end. Are they auto-generated?  What about the schema magic types?
* Should we offer a compile time tool to generate flat, clean types for schemas so that intellisense works well

If your up for it, could you try such a prototype. Freed from having to create clean, production code, you should be able to explore some concepts that will help guide us.

Thank you for offering

Michael

**Reply by mobsense** [reply]
Let me know if you want me to create a branch that you can commit to -- to enable us to share and look at the progress.

**Reply by rjmackay** [reply]
I'll have a play around in our fork. I'd probably lean toward keeping the existing structure to start with and just doing a 1-1 conversion. Once that's done, and tests continue to pass, then consider any clean up work.

I've tried a couple of the smaller files already and generating the .d.ts files "just works", so that's probably the best place to start.

**Reply by rjmackay** [reply]
Experiments so far: https://github.com/sensedeep/dynamodb-onetable/pull/494

**Reply by ghost** [reply]
it looks like this PR was closed. Is this feature still something you are interested in having / looking for contributors for?

**Comment by mobsense** [comment]
Yes, definitely. Need someone with strong TS experience to continue the prototype.
